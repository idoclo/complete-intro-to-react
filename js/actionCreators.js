// @flow

import axios from 'axios';
import { SET_SEARCH_TERM, ADD_API_DATA } from './actions';

export function setSearchTerm(searchTerm: string) {
  return { type: SET_SEARCH_TERM, payload: searchTerm };
}

export function addAPIData(apiData: Show) {
  return { type: ADD_API_DATA, payload: apiData };
}

// Now tackle problem of making asynchronous request inside a thunk
export function getAPIDetails(imdbID: string) {
  // return a thunk, ie a deferred action
  return (dispatch: Function) => {
    axios
      .get(`http://localhost:3000/${imdbID}`)
      .then(response => {
        dispatch(addAPIData(response.data));
      })
      .catch(error => {
        throw new Error('axios error', error); // eslint-disable-line no-console
      });
  }
}