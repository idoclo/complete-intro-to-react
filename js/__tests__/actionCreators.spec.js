// @flow

import moxios from 'moxios'; // a mock version of axios; it comes with the axios library
import { setSearchTerm, addAPIData, getAPIDetails } from '../actionCreators';

// Testing actionCreators is just snapshot testing

const masterOfNone = {
  title: "Master of None",
  year: "2015-",
  description: "The personal and professional life of Dev, a 30-year-old actor in New York.",
  poster: "mon.jpg",
  imdbID: "tt4635276",
  trailer: "6bFvb3WKISk",
  rating: "1.5"
}


test('setSearchTerm', () => {
  // an action creator function returns a JSON object so we can see if it matches a snapshot
  expect(setSearchTerm('Atlanta')).toMatchSnapshot();
});

test('addAPIData', () => {
  expect(addAPIData(masterOfNone)).toMatchSnapshot();
});

// testing thunk
// an integration test between getAPIDetails and addAPIData functions: testing that the correct api is being called and that the correct action is being dispatched
test('getAPIDetails', (done: Function) => {
  moxios.withMock(() => {
    const dispatchMock = jest.fn(); // a spy function
    const thunk = getAPIDetails(masterOfNone.imdbID);
    thunk(dispatchMock);
    moxios.wait(() => {
      const request = moxios.requests.mostRecent(); // 1st thing we test is: was it called with the right url / did our thunk reach out to the correct api?
      request
        .respondWith({
          status: 200,
          response: masterOfNone
        })
        .then(() => {
          expect(request.url).toEqual(`http://localhost:3000/${masterOfNone.imdbID}`);
          expect(dispatchMock).toBeCalledWith(addAPIData(masterOfNone)); // toBeCalledWith is a function of the spy function. 2nd thing we test is: is the correct action is being dispatched?
          done();
        });
    });
  });
});