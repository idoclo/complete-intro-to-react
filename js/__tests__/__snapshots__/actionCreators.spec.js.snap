// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addAPIData 1`] = `
Object {
  "payload": Object {
    "description": "The personal and professional life of Dev, a 30-year-old actor in New York.",
    "imdbID": "tt4635276",
    "poster": "mon.jpg",
    "rating": "1.5",
    "title": "Master of None",
    "trailer": "6bFvb3WKISk",
    "year": "2015-",
  },
  "type": "ADD_API_DATA",
}
`;

exports[`setSearchTerm 1`] = `
Object {
  "payload": "Atlanta",
  "type": "SET_SEARCH_TERM",
}
`;
